### PR TITLE
MQTT protocol over bluetooth

### PR DESCRIPTION
**Pull request by Tasm-Devil**
With this commit, the MQTT class now also supports Bluetooth. I wrote a new socket class based on [Adafruit_CircuitPython_ESP32SPI](https://github.com/adafruit/Adafruit_CircuitPython_ESP32SPI/blob/c28977f75e50cbfb6b8c5b3f7747d9dd23ffdfe6/adafruit_esp32spi/adafruit_esp32spi_socket.py). There is also an example. I split the MQTT class somewhat so that there is a clear separation between MQTT and the underlying protocol.

I hope this work will be useful to others.

Warm greetings from Germany.

### CODE REVIEW

File: adafruit_minimqtt.py [hunk 1]
```diff
@@ -46,7 +46,7 @@
 from micropython import const
 import adafruit_logging as logging
 
-__version__ = "0.0.0-auto.0"
+__version__ = "0.1.0-auto.0"
```
Comment (tannewt) [inline adafruit_minimqtt.py]: Leave this as-is. The build scripts replace it with the git tag.

**Comment by Tasm-Devil** [comment]
OK. It's ready to push.

**Comment by brentru** [comment]
Hi @Tasm-Devil  - Travis is failing here: https://travis-ci.com/adafruit/Adafruit_CircuitPython_MiniMQTT/builds/142054500#L299

Once this check passes, I'll review the PR.

**Comment by Tasm-Devil** [comment]
Can you point to the code? I can't see the bug.

**Comment by brentru** [comment]
@Tasm-Devil  There are a few PyLint errors with the code. These aren't bugs, they are syntax issues.

For example, on Line 301 of the attached URL: https://travis-ci.com/adafruit/Adafruit_CircuitPython_MiniMQTT/builds/142054500#L2301

```
************* Module adafruit_ble_socket
C:124, 0: Line too long (102/100) (line-too-long)
```

On line 124 of the `adafruit_ble_socket` module (here: https://github.com/adafruit/Adafruit_CircuitPython_MiniMQTT/pull/15/files#diff-2e0ccaa8f54e85ee63d7eadf896bd7cfR124), the line is 2 characters over the allowed limit (100 chars.). You could resolve this error by moving to the next line and indenting, or reducing the amount of characters on that line.

Please work through each error until they're resolved. Then, re-commit your code. We have a guide/reference for this on the Adafruit Learning System: https://learn.adafruit.com/improve-your-code-with-pylint/pylint-errors#working-through-pylint-errors-5-5

**Comment by Tasm-Devil** [comment]
fixed

File: .pylintrc [hunk 2]
```diff
@@ -52,7 +52,7 @@ confidence=
...
...parameter-unpacking,unpacking-in-except,old-raise-syntax,backtick,long-suffix,old-ne-operator,old-octal-literal,import-star-module-level,raw-checker-failed,bad-inline-option,locally-disabled,locally-enabled,file-ignored,suppressed-message,useless-suppression,deprecated-pragma,apply-builtin,basestring-builtin,buffer-builtin,cmp-builtin,coerce-builtin,execfile-builtin,file-builtin,long-builtin,raw_input-builtin,reduce-builtin,standarderror-builtin,unicode-builtin,xrange-builtin,coerce-method,delslice-method,getslice-method,setslice-method,no-absolute-import,old-division,dict-iter-method,dict-view-method,next-method-called,metaclass-assignment,indexing-exception,raising-string,reload-builtin,oct-method,hex-method,nonzero-method,cmp-method,input-builtin,round-builtin,intern-builtin,unichr-builtin,map-builtin-not-iterating,zip-builtin-not-iterating,range-builtin-not-iterating,filter-builtin-not-iterating,using-cmp-argument,eq-without-hash,div-method,idiv-method,rdiv-method,exception-message-attribute,invalid-str-codec,sys-max-int,bad-python3-import,deprecated-string-function,deprecated-str-translate-call,import-error,missing-docstring,invalid-name,bad-whitespace,consider-using-enumerate
```
Comment (brentru) [inline .pylintrc]: Please do not change `.pylintrc`. 

Instead, please fix the errors thrown by pylint or add a line to disable pylint checking (`# pylint: disable=some-message,another-one`) above the line which you feel is erroneous. I'll review the lines and be able to provide feedback.

File: adafruit_minimqtt.py [hunk 3]
```diff
@@ -92,7 +92,7 @@ class MQTT:
     :param int keep_alive: KeepAlive interval between the broker and the MiniMQTT client.
     """
     # pylint: disable=too-many-arguments,too-many-instance-attributes, not-callable, invalid-name, no-member
-    def __init__(self, socket, broker, port=None, username=None,
+    def __init__(self, socket, port=None, username=None,
```
Comment (brentru) [inline adafruit_minimqtt.py]: Why was the `broker` kwarg removed?

File: .pylintrc [hunk 4]
```diff
@@ -72,7 +72,7 @@ evaluation=10.0 - ((float(5 * error + warning + refactor + convention) / stateme
 
 # Template used to display messages. This is a python new-style format string
 # used to format the message information. See doc for all details
-#msg-template=
+msg-template='{path} {line}: {msg} ({symbol})'
```
Comment (brentru) [inline .pylintrc]: please do not change this line

File: .pylintrc [hunk 5]
```diff
@@ -156,7 +156,7 @@ ignored-classes=optparse.Values,thread._local,_thread._local
 # (useful for modules/projects where namespaces are manipulated during runtime
 # and thus existing member attributes cannot be deduced by static analysis. It
 # supports qualified module names, as well as Unix pattern matching.
-ignored-modules=board
+ignored-modules=
```
Comment (brentru) [inline .pylintrc]: please add back the `board` module

**Comment by brentru** [comment]
@Tasm-Devil  what hardware and code did you use to test this class?

**Comment by Tasm-Devil** [comment]
I tested it on the Feather nrf52840. The Software was my BLE2WebService site, you can find here: https://github.com/Tasm-Devil/BLE2WebService

**Comment by brentru** [comment]
Ok, will test with a CLUE board.

**Comment by Tasm-Devil** [comment]
Nice. I'd like to get my hands on one 🙂

**Comment by brentru** [comment]
Ok, where is `examples/adafruit_ble_socket.py` originally sourced from? You have ladyada as the author, but it's not an official adafruit repository.

If you made this module, please submit a pull request for it to the **Adafruit_CircuitPython_BLE** library (https://github.com/adafruit/Adafruit_CircuitPython_BLE). I feel it should be located there, ESP32SPI isn't part of MiniMQTT.

**Comment by tannewt** [comment]
Why do you need MQTTOverWifi and MQTTOverBluetooth classes? Can you encapsulate the wifi and bluetooth details so that the MQTT stuff doesn't need to know about it?

**Comment by Tasm-Devil** [comment]
@bentru The socket is borrowed from here: adafruit_esp32spi/adafruit_esp32spi_socket.py
I agree, that this socket-file should be located somewhere else and not in the examples folder. But I thought, temporarily it's ok.

**Comment by Tasm-Devil** [comment]
> Why do you need MQTTOverWifi and MQTTOverBluetooth classes? Can you encapsulate the wifi and bluetooth details so that the MQTT stuff doesn't need to know about it?

I need those two subclasses because they use different interfaces. I think, that the parent-class should only do the MQTT-stuff and the children-classes worry about the two interfaces and how to establish a network connection. You can see the differences in the constructors of these subclasses.

How should we proceed now?

**Comment by tannewt** [comment]
I'd much rather have one MQTT class that takes in a socket that encapsulates all of the transport (BLE vs WiFi) details. The details in the subclasses should be factored into socket.

@Tasm-Devil I realize this was the case before your PR but I'd love to see this cleaned up. Having everything done in socket will make it easier to use real socket and ethernet as we add it too.

@brentru do you have cycles to help with this?

**Comment by brentru** [comment]
@tannewt  I'll have cycles to help with this soon, need to learn BLE anyway :) 

Since we're splitting out Tasm-devil's ble socket implementation - where would the best place for it to go (Adafruit_CircuitPython_BLE? somewhere else? new library?)

**Comment by tannewt** [comment]
@brentru I think a new BLE_Socket library would be best. We'll want to move all of the Wifi stuff out of here as well and into the ESP32SPI socket. MQTT shouldn't need to worry about the transport details at all.

**Comment by Tasm-Devil** [comment]
> @brentru I think a new BLE_Socket library would be best. We'll want to move all of the Wifi stuff out of here as well and into the ESP32SPI socket. MQTT shouldn't need to worry about the transport details at all.

Sounds very good to me. This is how it should be done.

**Comment by kattni** [comment]
@Tasm-Devil Are you still working on this? We have made some changes to the library that will now require you to update this PR to resolve merge conflicts. Please let us know if you need assistance with that.

**Comment by Tasm-Devil** [comment]
Yes, I will update everything. But I will create a new Pull request when I am ready.
Is a new repo for the BLE_Socket library already created by adafruit or should I create a new repo and you fork it?

**Comment by Tasm-Devil** [comment]
@brentru I'm looking at your revision of the MQTT class. I think you unfortunately made the same assumption regarding the interface that [my interface](https://github.com/adafruit/Adafruit_CircuitPython_BLE/blob/master/adafruit_ble/services/nordic.py) cannot fulfill and why I had to create two subclasses.

You think of the interface class of [adafruit_wiznet5k.py](https://github.com/adafruit/Adafruit_CircuitPython_Wiznet5k/blob/master/adafruit_wiznet5k/adafruit_wiznet5k.py) or [adafruit_esp32spi.py](https://github.com/adafruit/Adafruit_CircuitPython_ESP32SPI/blob/master/adafruit_esp32spi/adafruit_esp32spi.py) 
These two classes have the method `unpretty_ip ()`, which my interface class [nordic.py](https://github.com/adafruit/Adafruit_CircuitPython_BLE/blob/master/adafruit_ble/services/nordic.py) does not have and does not need either.

This means that because of lines 123 and 240 my plan to use this class via WebSocket over Bluetooth fails.

**Comment by Tasm-Devil** [comment]
I think it's ready to merge now. Althought the 64B Buffer size of the Bluetooth Service class is quite small for more than one subscripted message, it works like a charm. Thanks to bentru for cleaning up that class.

**Comment by jposada202020** [comment]
@brentru and @tannewt could you take a look a this, I could work out with @Tasm-Devil the merge conflict in case this is good for you folks. Thanks

**Comment by tannewt** [comment]
I'd like to see it use the new socketpool/socket model. I think this library has been switched over. It feels like this could be a separate library too. What needs to be on the other end of the BLE connection? It seems weird to me that it uses a UART service.

**Comment by brentru** [comment]
I agree with Scott w.r.t updating + making it a sep library if anyone wants to take it on...but,

The more I think about it, I don't understand this PR's utility in relation to the library. The MQTT 3.1.1 protocol (which this library follows) transports over TCP/IP, but does not specify BLE as a transport [MQTT Specification, 4:2 Network Connections]:
```
The transport protocol used to carry MQTT 3.1 was TCP/IP as defined in [RFC793]. TCP/IP can
1207 be used for MQTT 3.1.1. The following are also suitable:

- 1208 TLS [RFC5246]
- 1209 WebSocket [R]
```

A different MQTT specification, MQTT-SN, defines a "virtual connection" which could be used with BLE. That would be another CircuitPython library, though.

**Comment by jposada202020** [comment]
@Tasm-Devil Let me know what yo think. But as we see it this would need to be in another library. let me know thanks

**Comment by Tasm-Devil** [comment]
The example is up to date now. The many changes done in the last year are so good, that my little example works very good with this refactored MQTT Library. I wouldn't put the adafruit_ble_socket.py in another library because it's only a little helper socket class which I consider as a quick fix for this special example.

**Comment by Tasm-Devil** [comment]
> I'd like to see it use the new socketpool/socket model. I think this library has been switched over. It feels like this could be a separate library too. What needs to be on the other end of the BLE connection? It seems weird to me that it uses a UART service.

[This](https://github.com/Tasm-Devil/BLE2WebSocket) is on the other side of the Bluetooth Connection. It will open an Secure Websocket to the broker. I want to use it in my class to inspect MQTT-Messages on the fly. The Page needs a little work but its cool to see the MQTT-Packages and try to understand them by reading the RFC.

**Comment by FoamyGuy** [comment]
Closing this for now. My understanding reading this chain is that we would want this functionality introduced in a separate library. 

@Tasm-Devil if you're interested in working on it still you could move the bluetooth implementation into it's own repo and we could add it to the community bundle if you like. If you do want to do that and need help you can ping me here or join the discord.

**Pull request closed without merging**